url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/ACS.java

```diff
@@ -0,0 +1,1521 @@
+	 * <p>
+	 * If you are using goals or specialized working memory chunks, remember that the 
+	 * condition of a rule that you want to address those chunks must contain nodes for 
+	 * any dimension-value pairs within those chunks that differ from the sensory information 
+	 * space.
+	 * @param R The collection of rules to add.
+	 */
+	public void addRules (Collection <? extends AbstractRule> R)
+	{
+		for(AbstractRule r : R)
+			addRule(r);
+	}
+	
+	/**
+	 * Generates a match all rule associated with the specified action and adds it to the list
+	 * of match all rules.
+	 * @param act The action for which a match all rule is to be generated.
+	 */
+	private void generateMatchAllRule (AbstractAction act)
+	{
+		GeneralizedConditionChunk c = new GeneralizedConditionChunk ();
+		for(Dimension d : InputSpace.values())
+		{
+			if(!d.containsKey(Drive.TypicalInputs.STIMULUS))
+				c.put(d.getID(),d.clone());
+		}
+		MatchAllRules.put(act.getID(), new RefineableRule (c,act));
+	}
+	
+	/**
+	 * Gets the match all rule associated with the specified action. If the action does
+	 * not have a match all rule, this method returns null. While it MAY be possible for
+	 * this method to return null, it SHOULD never happen.
+	 * @param act The action associated with the match all rule you wish to get.
+	 * @return The match all rule for the specified action.
+	 */
+	private AbstractRule getMatchAllRule (AbstractAction act)
+	{
+		return MatchAllRules.get(act.getID());
+	}
+	
+	/**
+	 * Updates the positive or negative match statistics for the match all rule corresponding 
+	 * to the chosen action.
+	 * @param ChosenAction The chosen action.
+	 * @param feedback The feedback.
+	 */
+	private void updateMatchAllStats(AbstractAction ChosenAction, double feedback, AbstractMatchCalculator MatchCalculator)
+	{
+		AbstractRule ma = MatchAllRules.get(ChosenAction.getID());
+		ma.setFeedback(feedback);
+		ma.updateMatchStatistics(MatchCalculator);
+	}
+	
+	/**
+	 * Performs the appropriate end of episode instructions for the ACS. This method is called by the
+	 * CLARION class when its endEpisode method is called.
+	 * @param Input A collection of various information to be used for ending the episode.
+	 * @param TimeStamp The current time stamp.
+	 */
+	protected void endEpisode (DimensionValueCollection Input, long TimeStamp)
+	{
+		RERRuleStore.discountMatchStatistics(MATCH_DISCOUNT);
+		IRLRuleStore.discountMatchStatistics(MATCH_DISCOUNT);
+		FRRuleStore.discountMatchStatistics(MATCH_DISCOUNT);
+		BLImplicitModuleStore.discountMatchStatistics();
+		
+		for(AbstractRule i : MatchAllRules.values())
+		{
+			i.setPM(i.getPM() * MATCH_DISCOUNT);
+			i.setNM(i.getNM() * MATCH_DISCOUNT);
+		}
+	}
+
+	/**
+	 * Attaches the ACS to the specified CLARION agent.
+	 * @param Agent The agent to wish this ACS will be attached.
+	 */
+	protected void attachSelfToAgent(CLARION Agent) {
+		Agent.attachACS(this);
+	}
+	
+	protected void updateInputSpace (Collection <Dimension> c)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 27. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.